### PR TITLE
fix(stream): Better cache handling for temporal join.

### PR DESCRIPTION
**Pull request by wsx-ucb**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Attempt to evict cache and update vnodes on every epoch for the temporal join executor.

## Checklist For Contributors

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [x] I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer to the issue)
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Checklist For Reviewers

- [ ] I have requested macro/micro-benchmarks as this PR can affect performance substantially, and the results are shown.

## Documentation

- [x] My PR **DOES NOT** contain user-facing changes.

### Release note

</details>

### CODE REVIEW

File: src/stream/src/executor/temporal_join.rs [hunk 1]
```diff
@@ -195,7 +195,11 @@ impl<S: StateStore, const T: JoinTypePrimitive> TemporalJoinExecutor<S, T> {
 
         let alloc = StatsAlloc::new(Global).shared();
 
-        let cache = new_with_hasher_in(watermark_epoch, DefaultHasher::default(), alloc);
+        let cache = ExecutorCache::new(new_with_hasher_in(
+            watermark_epoch,
+            DefaultHasher::default(),
+            alloc,
```
Comment (BugenZhao) [inline src/stream/src/executor/temporal_join.rs:201]: Seems using the default `Global` allocator is enough? cc @yuhao-su

File: src/stream/src/executor/temporal_join.rs [hunk 2]
```diff
@@ -277,9 +281,18 @@ impl<S: StateStore, const T: JoinTypePrimitive> TemporalJoinExecutor<S, T> {
...
-                    prev_epoch = Some(barrier.epoch.curr);
+                    self.right_table.cache.evict();
+                    if let Some(vnodes) = barrier.as_update_vnode_bitmap(self.ctx.id) {
+                        let prev_vnodes =
+                            self.right_table.source.update_vnode_bitmap(vnodes.clone());
+                        if cache_may_stale(&prev_vnodes, &vnodes) {
+                            self.right_table.cache.clear();
+                        }
+                    }
+                    self.right_table.cache.update_epoch(barrier.epoch.curr);
```
Comment (BugenZhao) [inline src/stream/src/executor/temporal_join.rs:292]: Is this already called in `right_table.update`?

File: src/stream/src/executor/temporal_join.rs [hunk 3]
```diff
@@ -59,7 +59,7 @@ pub struct TemporalJoinExecutor<S: StateStore, const T: JoinTypePrimitive> {
 struct TemporalSide<S: StateStore> {
     source: StorageTable<S>,
     table_output_indices: Vec<usize>,
-    cache: ManagedLruCache<OwnedRow, Option<OwnedRow>, DefaultHasher, SharedStatsAlloc<Global>>,
+    cache: ExecutorCache<OwnedRow, Option<OwnedRow>, DefaultHasher, SharedStatsAlloc<Global>>,
```
Comment (BugenZhao) [inline src/stream/src/executor/temporal_join.rs:62]: What does `None` in the value mean here?

File: src/stream/src/executor/temporal_join.rs (hunk 2 above)
Comment (wsx-ucb) [reply src/stream/src/executor/temporal_join.rs:292]: Thanks for pointing this out! I meant to only call `update_epoch` in the `into_stream` implementation, and I'll remove the other occurrence.

**Review by chenzl25 — APPROVED** [review]
LGTM

**Pull request merged**